### PR TITLE
Group20 to python 3

### LINKED ISSUE #19821: Group 20 to python 3
Group 20 (below) of #18550 are to be refactored to be both python 2 and 3 compatible. 

**Group 20**
```
scripts/HFIR_Powder_Diffraction_Reduction.py
scripts/HFIRPowderReduction/HfirUtility.py
scripts/HFIRPowderReduction/HfirPDReductionGUI.py
scripts/HFIRPowderReduction/MplFigureCanvas.py
scripts/HFIRPowderReduction/__init__.py
scripts/HFIRPowderReduction/HfirPDReductionControl.py
```

### PR DESCRIPTION
**Pull request by AnthonyLim23**
This is the conversion of group 20 into python 3. This is a simple refactoring of the code. It should be compatible with both python 2 and 3. 

Fixes #19821 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by wdzhou — CHANGES_REQUESTED** [review]
There are conflicts in 'scripts/HFIRPowderReduction/HfirPDReductionControl.py'.

I am going to resolve this.

**Review by wdzhou — APPROVED** [review]

**Comment by wdzhou** [comment]
All changes look good!  :shipit:

File: scripts/HFIRPowderReduction/HfirPDReductionGUI.py [hunk 1]
```diff
@@ -72,7 +79,7 @@ def compare_state(self, tab_state):
         if self._expNo != tab_state.getExpNumber() or self._scanList != tab_state.getScanList:
             return self.RELOAD_DATA
 
-        for attname in self.__dict__.keys():
+        for attname in list(self.__dict__.keys()):
```
Comment (rosswhitfield) [inline scripts/HFIRPowderReduction/HfirPDReductionGUI.py]: You don't need to convert to list if you are iterating over it.

File: scripts/HFIRPowderReduction/HfirPDReductionGUI.py [hunk 2]
```diff
@@ -2510,12 +2517,12 @@ def _getIntArray(self, intliststring):
                     if returnstatus is False:
                         break
                 # ENDFOR
-                intlist.extend(range(templist[0], templist[1]+1))
+                intlist.extend(list(range(templist[0], templist[1]+1)))
```
Comment (rosswhitfield) [inline scripts/HFIRPowderReduction/HfirPDReductionGUI.py]: Don't need list here either

File: scripts/HFIRPowderReduction/MplFigureCanvas.py [hunk 3]
```diff
@@ -419,16 +420,16 @@ def addImage(self, imagefilename):
     def clearAllLines(self):
         """ Remove all lines from the canvas
         """
-        for ikey in self._lineDict.keys():
+        for ikey in list(self._lineDict.keys()):
```
Comment (rosswhitfield) [inline scripts/HFIRPowderReduction/MplFigureCanvas.py]: and here

**Review by rosswhitfield — CHANGES_REQUESTED** [review]

**Review by rosswhitfield — APPROVED** [review]

**Review by rosswhitfield — APPROVED** [review]

**Review by rosswhitfield — APPROVED** [review]

**Review by rosswhitfield — APPROVED** [review]

**Pull request merged**